### ISSUE TITLE
Running `check` Command Too Soon After Stop Reports Damaged Database

### ISSUE THREAD

**Issue by SE2Dev** [ChuckPa/DBRepair#144]
In cases where the `check` command is run very shortly after the `stop` command, the main database will be reported as damaged:
```
Checking the PMS databases
Killed
Check complete.  PMS main database is damaged.
Check complete.  PMS blobs database is OK.
```

The `Killed` line is particularly noteworthy as it only seems to show up when this issue is occurring.

Additionally running the `automatic` command instead of explicitly running `check` can result in additional weirdness:
```log
Enter command # -or- command name (4 char min) : automatic
 
Automatic Check,Repair,Index started.

Checking the PMS databases
Killed
Check complete.  PMS main database is damaged.
Check complete.  PMS blobs database is OK.

Exporting current databases using timestamp: 2024-04-28_18.39.11
Exporting Main DB
Killed
Error 137 from Plex SQLite while exporting com.plexapp.plugins.library.db
Could not successfully export the main database to repair it.  Please try restoring a backup.
Repair failed. Automatic mode cannot continue. Please repair with individual commands
```

Upon attempting to run `automatic` the first time, and having it fail, all subsequent calls to `check` and `automatic` trigger the same issue(s), and requires manual intervention to restore the database (as `replace` won't work either)
```
Enter command # -or- command name (4 char min) : replace
 
Checking the PMS databases
Killed
Check complete.  PMS main database is damaged.
Check complete.  PMS blobs database is OK.
Checking for a usable backup.
Database backups available are:  2024-04-26 2024-04-23 2024-04-20 2024-04-17
Checking database 2024-04-26
Killed
Checking database 2024-04-23
Killed
Checking database 2024-04-20
Killed
Checking database 2024-04-17
Killed
Error.  No valid matching main and blobs database pairs.  Cannot replace.
```

**Comment by ChuckPa** [comment]
Need to know the details of the environment this is running in.
(Distro & whether a container, VM or native)

The "Killed" line is because `Plex SQLite` is being killed.
Getting killed will return a failed status code and be interpreted as "Damaged DB"  ( $status != 0 )

**Comment by SE2Dev** [comment]
Oops - sorry, it's the standard Docker image.

Somewhere during the process of writing this post I seem to have caused my database to become _actually_ corrupt, so now I'm working through the process of restoring it.

**Comment by ChuckPa** [comment]
I'll be around.

There have never been reported failures of Plex SQLite in my script unless there are corresponding failures in PMS.

Please let me know what you find out ?

**Comment by SE2Dev** [comment]
Even after restoring the backup I made (before starting PlexDBRepair.sh) the first time, I'm still getting the following error upon restarting the container:
```
plex  | Error: Unable to set up server: sqlite3_statement_backend::loadOne: attempt to write a readonly database (N4soci10soci_errorE)
plex  | Stopping Plex Media Server.
plex  | kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec
```

I did double check the permissions for `com.plexapp.plugins.library.db` and they looked good.

**Comment by ChuckPa** [comment]
good

**Comment by SE2Dev** [comment]
Yes, the initial issue I mentioned still occurs though.
```
Select

  1 - 'stop'      - Stop PMS.
  2 - 'automatic' - Check, Repair/Optimize, and Reindex Database in one step.
  3 - 'check'     - Perform integrity check of database.
  4 - 'vacuum'    - Remove empty space from database without optimizing.
  5 - 'repair'    - Repair/Optimize databases.
  6 - 'reindex'   - Rebuild database database indexes.
  7 - 'start'     - Start PMS

  8 - 'import'    - Import watch history from another database independent of Plex. (risky).
  9 - 'replace'   - Replace current databases with newest usable backup copy (interactive).
 10 - 'show'      - Show logfile.
 11 - 'status'    - Report status of PMS (run-state and databases).
 12 - 'undo'      - Undo last successful command.

 21 - 'prune'     - Prune (remove) old image files (jpeg,jpg,png) from PhotoTranscoder cache.
 42 - 'ignore'    - Ignore duplicate/constraint errors.

 88 - 'update'    - Check for updates.
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) : stop
 
Stopping PMS.
Stopped PMS.

Select

  1 - 'stop'      - Stop PMS.
  2 - 'automatic' - Check, Repair/Optimize, and Reindex Database in one step.
  3 - 'check'     - Perform integrity check of database.
  4 - 'vacuum'    - Remove empty space from database without optimizing.
  5 - 'repair'    - Repair/Optimize databases.
  6 - 'reindex'   - Rebuild database database indexes.
  7 - 'start'     - Start PMS

  8 - 'import'    - Import watch history from another database independent of Plex. (risky).
  9 - 'replace'   - Replace current databases with newest usable backup copy (interactive).
 10 - 'show'      - Show logfile.
 11 - 'status'    - Report status of PMS (run-state and databases).
 12 - 'undo'      - Undo last successful command.

 21 - 'prune'     - Prune (remove) old image files (jpeg,jpg,png) from PhotoTranscoder cache.
 42 - 'ignore'    - Ignore duplicate/constraint errors.

 88 - 'update'    - Check for updates.
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) : check
 
Checking the PMS databases
Killed
Check complete.  PMS main database is damaged.
Check complete.  PMS blobs database is OK.

Select

  1 - 'stop'      - Stop PMS.
  2 - 'automatic' - Check, Repair/Optimize, and Reindex Database in one step.
  3 - 'check'     - Perform integrity check of database.
  4 - 'vacuum'    - Remove empty space from database without optimizing.
  5 - 'repair'    - Repair/Optimize databases.
  6 - 'reindex'   - Rebuild database database indexes.
  7 - 'start'     - Start PMS

  8 - 'import'    - Import watch history from another database independent of Plex. (risky).
  9 - 'replace'   - Replace current databases with newest usable backup copy (interactive).
 10 - 'show'      - Show logfile.
 11 - 'status'    - Report status of PMS (run-state and databases).
 12 - 'undo'      - Undo last successful command.

 21 - 'prune'     - Prune (remove) old image files (jpeg,jpg,png) from PhotoTranscoder cache.
 42 - 'ignore'    - Ignore duplicate/constraint errors.

 88 - 'update'    - Check for updates.
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) : check
 
Checking the PMS databases
Check complete.  PMS main database is OK.
Check complete.  PMS blobs database is OK.

Select

  1 - 'stop'      - Stop PMS.
  2 - 'automatic' - Check, Repair/Optimize, and Reindex Database in one step.
  3 - 'check'     - Perform integrity check of database.
  4 - 'vacuum'    - Remove empty space from database without optimizing.
  5 - 'repair'    - Repair/Optimize databases.
  6 - 'reindex'   - Rebuild database database indexes.
  7 - 'start'     - Start PMS

  8 - 'import'    - Import watch history from another database independent of Plex. (risky).
  9 - 'replace'   - Replace current databases with newest usable backup copy (interactive).
 10 - 'show'      - Show logfile.
 11 - 'status'    - Report status of PMS (run-state and databases).
 12 - 'undo'      - Undo last successful command.

 21 - 'prune'     - Prune (remove) old image files (jpeg,jpg,png) from PhotoTranscoder cache.
 42 - 'ignore'    - Ignore duplicate/constraint errors.

 88 - 'update'    - Check for updates.
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) : 
```

**Comment by ChuckPa** [comment]
The test for whether PMS is running or not  occurs BEFORE the check starts.
```
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) : check
 
Checking the PMS databases
Killed
Check complete.  PMS main database is damaged.
Check complete.  PMS blobs database is OK.

Select```

Something in your environment is still killing -- OR -- Restarting because of Health check failing.

Whose image are you using?
Which host ?

**Comment by SE2Dev** [comment]
Image: `plexinc/pms-docker:latest` (currently running PMS 1.40.2.8395)
Host: Synology running Linux kernel `4.4.302+ `

**Comment by SE2Dev** [comment]
The container itself is run via Docker Compose, but the DBRepair script is being run by attaching to the container shell.

**Comment by ChuckPa** [comment]
I will need walk this through the your steps.

Since HDR became standard on DSM,    I don't know who still uses docker on the machines.
(cutting out Docker removes more overhead from those limited CPUs)

**Comment by SE2Dev** [comment]
> Never known anyone to try things with Docker attach. Most do a Docker exec
Sorry, yes - I attached to the container via the Docker extension in VSCode ("Attach Shell"), but it looks like it's literally just running `docker exec -it [container id] bash`. At which point I would do:
```shell
cd /tmp
apt-get update && apt-get install -y wget
wget https://github.com/ChuckPa/PlexDBRepair/releases/download/v1.05.02/DBRepair.sh
chmod +x ./DBRepair.sh
./DBRepair.sh
```

> Since HDR became standard on DSM, I don't know who still uses docker on the machines. (cutting out Docker removes more overhead from those limited CPUs)
I'm not really sure what you mean by HDR, are you talking about support for HDR media in Plex via the native Plex Media Server package?

**Comment by ChuckPa** [comment]
1.  "HDR"  ==  HEVC HDR with tonemappping support as native.
2.  I'm also the Plex engineer who supports Synology and QNAP systems

Also,  be advised,   DBRepair  has its own "update" capability.   You no longer need to update manually.


I'll setup a container and see what happens.

**Comment by SE2Dev** [comment]
> I'm also the Plex engineer who supports Synology and QNAP systems
That's probably one of the many places I recognize you from.

For what it's worth, I did just confirm that this does happen running the same container on a regular Debian host. You just seem to need a somewhat sizable database size (my blobs and main databases are about 500mb each).;

**Comment by ChuckPa** [comment]
I just ran this on 1GB databases without issue.

I will do my best to figure out why it's  (Plex SQLite) failing for you.

This might be entirely a Plex issue with the latest version (would not be surprised)

**Comment by SE2Dev** [comment]
I can confirm that this has been happening for at least the last few "released to everyone" versions, as I originally used this project to verify the database after one of the recent updates triggered migrations that required several steps to revert (just to make sure that the upgrade had worked properly). (I had seen some spooky console output during the server upgrade & "Optimize Database" run, and wanted to make sure my database wasn't corrupt).

**Comment by ChuckPa** [comment]
1.40.0.7775 is where Engineering released an entirely new database schema.
There was a HUGE schema migration time required.

Folks who kept restarting before it finished ended up with tons of problems.

I'm in the process of standing up a 2.6 GB DB and will let it rebuild the entire server from the DB file.
If it doesn't fail there then the issue must be in the health check / container config.

Speaking of that,   do you impose memory limits in your container ?

**Comment by SE2Dev** [comment]
> Speaking of that, do you impose memory limits in your container ?
Nope (Debian system I reproduced the issue on has 64GB of memory)

**Comment by ChuckPa** [comment]
Not "have",  do you impose limits in the container startup?
(the container's definition)

As for 64 GB in a Syno ---  Is the an XPEnology machine?

**Comment by ChuckPa** [comment]
Just reproduced it.

Digging into Plex's  `Plex SQLite` now.

```
root@dockerplex:~/Library/Application Support/Plex Media Server/Plug-in Support/Databases# ./DBRepair.sh stop auto start exit
 
 
 
      Plex Media Server Database Repair Utility (Docker)
                       Version v1.05.02
 
 
[2024-04-28 20.53.23] Stopping PMS.
[2024-04-28 20.53.26] Stopped PMS.
 
[2024-04-28 20.53.26] Automatic Check,Repair,Index started.
[2024-04-28 20.53.26] 
[2024-04-28 20.53.26] Checking the PMS databases
Killed
[2024-04-28 20.53.28] Check complete.  PMS main database is damaged.
[2024-04-28 20.53.28] Check complete.  PMS blobs database is OK.
[2024-04-28 20.53.28] 
[2024-04-28 20.53.28] Exporting current databases using timestamp: 2024-04-28_20.53.26
[2024-04-28 20.53.28] Exporting Main DB
^C
```

**Comment by SE2Dev** [comment]
> Not "have", do you impose limits in the container startup? (the container's definition)

Like I said, no - there's no limits (of any type) being specified in the `docker-compose.yaml`.

> How much memory is in the machine?

The Synology machine that I originally saw the issue on has 8GB, the Debian machine that I reproduced it on later had 64GB.

**Comment by ChuckPa** [comment]
I'm using their latest image right now.    I think that's where the problem is -- their current executable.

**Comment by ChuckPa** [comment]
Let's see if this is a workaround and constrains the problem.

DBRepair.sh:  

At Line 1401, insert -

```
echo Sleeping for Plex workaround
sleep 10
```


If my testing is right, they are holding onto something in their new "One process to rule all" code which is flawed.


Ref:
```
   1393     $StopCommand > /dev/null 2> /dev/null
   1394     Result=$?
   1395     if [ $Result -ne 0 ]; then
   1396       Output   "Cannot send stop command to PMS, error $Result.  Please stop manually."
   1397       WriteLog "Cannot send stop command to PMS, error $Result.  Please stop manually."
   1398       return 1
   1399     fi
   1400 
   1401 echo sleeping 10 seconds
   1402 sleep 10
   1403 
   1404     Count=10
   1405     while IsRunning && [ $Count -gt 0 ]
   1406     do
   1407       sleep 3
   1408       Count=$((Count - 1))
   1409     done
   1410 
```

**Comment by SE2Dev** [comment]
That *does* seem to serve as a workaround for the time being, although it might be worth pointing out that for whatever reason that seems to be line 1512, not 1401. With that added, the extra time seems to be enough for the "Plex SQLite" instance to wrap up whatever it's doing.

**Comment by ChuckPa** [comment]
I might be off by a few lines .. That's why I gave you the code block to look at.

I am waiting for Plex to tell me what changed and why it's so screwy.

**Comment by ChuckPa** [comment]
I am putting together something a little more reasonable than that hack of a delay.

How about Manual mode?
Something you can invoke OUTSIDE the container if you can specify the paths 

This is first cut,  cleanup is needed.

```
root@lizum:/home/chuck/git/chuck/PlexDBRepair# ./DBRepair.sh --sqlite /usr/lib/plexmediaserver --databases "/usb/plex/Plug-in Support/Databases" -p
==== Setting DBDIR = "/usb/plex/Plug-in Support/Databases"
 
 
 
      Plex Media Server Database Repair Utility ()
                       Version v1.06.00
 
      PlexSQLite = '/usr/lib/plexmediaserver/Plex SQLite'
      Databases  = '/usb/plex/Plug-in Support/Databases'

Select

  1 - 'stop'      - (Not available. Stop manually.)
  2 - 'automatic' - Check, Repair/Optimize, and Reindex Database in one step.
  3 - 'check'     - Perform integrity check of database.
  4 - 'vacuum'    - Remove empty space from database without optimizing.
  5 - 'repair'    - Repair/Optimize databases.
  6 - 'reindex'   - Rebuild database database indexes.
  7 - 'start'     - (Not available. Start manually)

  8 - 'import'    - Import watch history from another database independent of Plex. (risky).
  9 - 'replace'   - Replace current databases with newest usable backup copy (interactive).
 10 - 'show'      - Show logfile.
 11 - 'status'    - Report status of PMS (run-state and databases).
 12 - 'undo'      - Undo last successful command.

 21 - 'prune'     - Prune (remove) old image files (jpeg,jpg,png) from PhotoTranscoder cache.
 42 - 'ignore'    - Ignore duplicate/constraint errors.

 88 - 'update'    - Check for updates.
 99 - 'quit'      - Quit immediately.  Keep all temporary files.
      'exit'      - Exit with cleanup options.

Enter command # -or- command name (4 char min) :```

**Comment by ChuckPa** [comment]
Update:    This is verified as a PMS bug.

I've chatted with the engineer and submitted a ticket.

I'm closing this for now but will be watching my PMS ticket to be resolved.

In the interim,  add as much sleep after the stop as is needed to satisfy their new "one process for all" mechanism.
(The processes used to be distinct but now have been merged into one executable)

**Comment by ChuckPa** [comment]
@johnny2678 

Are you attempting to deflate a bloated DB?

What does `ls -lah` show for the sizes of your databases ?

**Comment by johnny2678** [comment]
Hi @ChuckPa - thanks for the quick response.  Here's the `la-lah` output:

```
bash: la: command not found
root@15620-BEAST:~/Library/Application Support/Plex Media Server/Plug-in Support/Databases# ls -lah
total 3.4G
drwxr-xr-x 1 plex users 2.4K Jul  7 12:16 .
drwxr-xr-x 1 plex users  100 Aug 22  2020 ..
drwxrwxrwx 1 root root   136 Jul  7 09:50 DBRepair
-rw-r--r-- 1 root root   11K Jul  7 12:16 DBRepair.log
-rw-rw-rw- 1 root root  709M Jul  7 09:45 bak.tar
-rw-r--r-- 1 plex users 432M Jul  6 02:20 com.plexapp.plugins.library.blobs.db
-rw-r--r-- 1 plex users 395M Jun 29 02:26 com.plexapp.plugins.library.blobs.db-2025-06-29
-rw-r--r-- 1 plex users  32K Jul  7 11:50 com.plexapp.plugins.library.blobs.db-2025-06-29-shm
-rw-r--r-- 1 plex users    0 Jul  7 10:00 com.plexapp.plugins.library.blobs.db-2025-06-29-wal
-rw-r--r-- 1 plex users 396M Jul  2 02:57 com.plexapp.plugins.library.blobs.db-2025-07-02
-rw-r--r-- 1 plex users  32K Jul  7 09:59 com.plexapp.plugins.library.blobs.db-2025-07-02-shm
-rw-r--r-- 1 plex users    0 Jul  7 09:59 com.plexapp.plugins.library.blobs.db-2025-07-02-wal
-rw-r--r-- 1 plex users 392M Jul  5 02:26 com.plexapp.plugins.library.blobs.db-2025-07-05
-rw-r--r-- 1 plex users  32K Jul  7 09:58 com.plexapp.plugins.library.blobs.db-2025-07-05-shm
-rw-r--r-- 1 plex users    0 Jul  7 09:58 com.plexapp.plugins.library.blobs.db-2025-07-05-wal
-rw-r--r-- 1 plex users 278M Jul  7 07:32 com.plexapp.plugins.library.db
-rw-r--r-- 1 plex users 257M Jun 29 02:26 com.plexapp.plugins.library.db-2025-06-29
-rw-r--r-- 1 plex users  32K Jul  7 11:49 com.plexapp.plugins.library.db-2025-06-29-shm
-rw-r--r-- 1 plex users    0 Jul  7 10:00 com.plexapp.plugins.library.db-2025-06-29-wal
-rw-r--r-- 1 plex users 279M Jul  2 02:56 com.plexapp.plugins.library.db-2025-07-02
-rw-r--r-- 1 plex users  32K Jul  7 09:59 com.plexapp.plugins.library.db-2025-07-02-shm
-rw-r--r-- 1 plex users    0 Jul  7 09:59 com.plexapp.plugins.library.db-2025-07-02-wal
-rw-r--r-- 1 plex users 277M Jul  5 02:26 com.plexapp.plugins.library.db-2025-07-05
-rw-r--r-- 1 plex users  32K Jul  7 09:58 com.plexapp.plugins.library.db-2025-07-05-shm
-rw-r--r-- 1 plex users    0 Jul  7 09:58 com.plexapp.plugins.library.db-2025-07-05-wal
-rw-r--r-- 1 plex users  32K Jul  7 13:06 com.plexapp.plugins.library.db-shm
-rw-r--r-- 1 plex users    0 Jul  7 12:14 com.plexapp.plugins.library.db-wal
drwxr-xr-x 1 root root    20 Jul  7 11:34 manual_backups
-rw-r--r-- 1 plex users  67M Jul  7 07:31 tv.plex.providers.epg.cloud-13f4e99d-0962-4bd6-bd0a-aceae2242119.db
-rw-r--r-- 1 plex users  32K Jul  7 07:31 tv.plex.providers.epg.cloud-13f4e99d-0962-4bd6-bd0a-aceae2242119.db-shm
-rw-r--r-- 1 plex users 1.1M Jul  7 07:31 tv.plex.providers.epg.cloud-13f4e99d-0962-4bd6-bd0a-aceae2242119.db-wal
```

Plex was working fine yesterday.  Today, I went to put on a show for my 3 year old and noticed all libraries were unavailable.  Docker logs show the following sqlite3 errors on repeat:
```
[cont-init.d] executing container initialization scripts...
[cont-init.d] 40-plex-first-run: executing... 
[cont-init.d] 40-plex-first-run: exited 0.
[cont-init.d] 45-plex-hw-transcode-and-connected-tuner: executing... 
[cont-init.d] 45-plex-hw-transcode-and-connected-tuner: exited 0.
[cont-init.d] 50-plex-update: executing... 
[cont-init.d] 50-plex-update: exited 0.
[cont-init.d] done.
[services.d] starting services
Starting Plex Media Server.
[services.d] done.
Error: Unable to set up server: sqlite3_statement_backend::loadOne: database disk image is malformed (N4soci10soci_errorE)
Stopping Plex Media Server.
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
Starting Plex Media Server.
Error: Unable to set up server: sqlite3_statement_backend::loadOne: database disk image is malformed (N4soci10soci_errorE)
Stopping Plex Media Server.
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]

```

I've tried a number of DBRepair script options but all of them generate some version of the sqlite errors above.  And I can't seem to restore, either from Plex automated backups, or from my borgmatic backups that go back 1 year.  Has my DB been corrupt for that long??

**Comment by ChuckPa** [comment]
I  am seeing this on proper shutdown docker containers.

I am reopening this to conduct research.

Given I use `Plex SQLite`,  I suspect they have a regression or new bug in PMS shutdown.

**Comment by ChuckPa** [comment]
When you next have PMS stopped,

Please delete all the -wal and -shm files which are associated with those backup databases.
They serve no purpose.

**Comment by johnny2678** [comment]
> Please delete all the -wal and -shm files which are associated with those backup databases.
> They serve no purpose.

Done.

My database was damaged again 24 hours after my previous restore. blobs was ok.

```
Error: Unable to set up server: sqlite3_statement_backend::loadOne: database disk image is malformed (N4soci10soci_errorE)
Stopping Plex Media Server.
```

this time, I was able to use DBRepair to `9 - replace` from a recent backup and plex is running again.

twice in 48 hours after running plex for 10+ years.  I guess I need to up my backup frequency.

**Comment by Drakonas** [comment]
I am getting this issue again. I've found it to be caused by an issue where the host keeps installing the plexmediaserver package, which prevents plex from starting up in the docker container. It does this even though I've removed the package. I found out I forgot I had installed the [plexupdate](https://github.com/mrworf/plexupdate) cron job which automatically installs it on the host.

Basically, if the host has plexmediaserver running, but you run the script in the docker container, these errors come up that are listed in this issue.

If you have plexupdate. Remove the job file from /etc/cron.daily and /etc/cron.weekly (whichever exists) and then remove the /etc/plexupdate*.conf files (there were two for me).
